### ISSUE TITLE
cannot compile with gcc-5.4

### ISSUE THREAD

**Issue by asartori86** [dealii/dealii#4661]
just pulled the master branch and I cannot compile it anymore with gcc-5.4. Is this a bug in my stl?
```
[ 67%] Building CXX object source/base/CMakeFiles/obj_base_release.dir/function_parser.cc.o
In file included from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_tempbuf.h:60:0,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_algo.h:62,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/algorithm:62,
                 from /scratch/opt/src/dealii/include/deal.II/base/table_indices.h:23,
                 from /scratch/opt/src/dealii/include/deal.II/base/tensor.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/auto_derivative_function.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function_parser.h:22,
                 from /scratch/opt/src/dealii/source/base/function_parser.cc:17:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_construct.h: In instantiation of ‘void std::_Construct(_T1*, _Args&& ...) [with _T1 = std::unique_ptr<mu::Parser>; _Args = {const std::unique_ptr<mu::Parser, std::default_delete<mu::Parser> >&}]’:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:75:18:   required from ‘static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*; bool _TrivialValueTypes = false]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:126:15:   required from ‘_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_uninitialized.h:281:37:   required from ‘_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::unique_ptr<mu::Parser>*, std::vector<std::unique_ptr<mu::Parser> > >; _ForwardIterator = std::unique_ptr<mu::Parser>*; _Tp = std::unique_ptr<mu::Parser>]’
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_vector.h:322:31:   required from ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = std::unique_ptr<mu::Parser>; _Alloc = std::allocator<std::unique_ptr<mu::Parser> >]’
/usr/include/tbb/enumerable_thread_specific.h:477:49:   required from ‘tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::padded_element<U>::padded_element(const U&) [with U = std::vector<std::unique_ptr<mu::Parser> >; T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u]’
/usr/include/tbb/enumerable_thread_specific.h:636:91:   required from ‘T& tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::local(bool&) [with T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u; tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::reference = std::vector<std::unique_ptr<mu::Parser> >&]’
/usr/include/tbb/enumerable_thread_specific.h:602:25:   required from ‘T& tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::local() [with T = std::vector<std::unique_ptr<mu::Parser> >; Allocator = tbb::cache_aligned_allocator<std::vector<std::unique_ptr<mu::Parser> > >; tbb::ets_key_usage_type ETS_key_type = (tbb::ets_key_usage_type)1u; tbb::enumerable_thread_specific<T, Allocator, ETS_key_type>::reference = std::vector<std::unique_ptr<mu::Parser> >&]’
/scratch/opt/src/dealii/include/deal.II/base/thread_local_storage.h:210:23:   required from ‘T& dealii::Threads::ThreadLocalStorage<T>::get() [with T = std::vector<std::unique_ptr<mu::Parser> >]’
/scratch/opt/src/dealii/source/base/function_parser.cc:383:13:   required from ‘double dealii::FunctionParser<dim>::value(const dealii::Point<dim>&, unsigned int) const [with int dim = 1]’
/scratch/opt/src/dealii/source/base/function_parser.cc:478:16:   required from here
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/stl_construct.h:75:7: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = mu::Parser; _Dp = std::default_delete<mu::Parser>]’
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^
In file included from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/locale_conv.h:41:0,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/locale:43,
                 from /u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/iomanip:43,
                 from /scratch/opt/lib/trilinos/include/Epetra_ConfigDefs.h:152,
                 from /scratch/opt/lib/trilinos/include/Epetra_Comm.h:47,
                 from /scratch/opt/src/dealii/include/deal.II/base/utilities.h:28,
                 from /scratch/opt/src/dealii/include/deal.II/base/tensor.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/function.h:24,
                 from /scratch/opt/src/dealii/include/deal.II/base/auto_derivative_function.h:21,
                 from /scratch/opt/src/dealii/include/deal.II/base/function_parser.h:22,
                 from /scratch/opt/src/dealii/source/base/function_parser.cc:17:
/u/shared/programs/x86_64/gcc/5.4.0/include/c++/5.4.0/bits/unique_ptr.h:356:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
make[2]: *** [source/base/CMakeFiles/obj_base_release.dir/function_parser.cc.o] Error 1
make[1]: *** [source/base/CMakeFiles/obj_base_release.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by asartori86** [comment]
if we change line 334 of `function_parser.h` to use `std::shared_ptr` instead of `std::unique_ptr` everything goes fine. 

I guess the issue with the current implementation is line 238 of `function_parser.cc` but I could not find a way to make it working with `std::unique_ptr`

**Comment by drwells** [comment]
I suspect a simpler implementation would make GCC happy: I am surprised that things work on 4.8, 6.3, and 7.1 but not here. Could you please try including the header

```cpp
#include <deal.II/base/std_cxx14/memory.h>
```
and then changing line 238 to
```cpp
fp.get().push_back(std_cxx14::make_unique<mu::Parser>());
```

**Comment by asartori86** [comment]
thanks @drwells, already tried but does not work..

the only solution I came up with is using `shared_ptr`.

**Comment by drwells** [comment]
Hm. It sounds like we should just go with `shared_ptr` then (these objects are only created once so there really isn't any extra expense). Would you mind writing a patch and adding a comment reminding our future selves why we did this?

Out of curiousity: which versions of TBB and muparser are you using?

**Comment by asartori86** [comment]
ok patch incoming

I use the bundled muparser. for tbb I got this info
```
$ yum info tbb
Installed Packages
Name        : tbb
Arch        : x86_64
Version     : 2.2
Release     : 3.20090809.el6
Size        : 201 k
Repo        : installed
From repo   : anaconda-SISSALinux-201605131214.x86_64
```

**Comment by Rombur** [comment]
TBB 2.2 is very old. The latest version is 4.4

**Comment by masterleinad** [comment]
Actually, I would really like to understand what is going on. I can't reproduce with gcc-5.4.0. The errors suggest that `mu::Parser` doesn't have a move constructor. But I fail to see why.

**Comment by davydden** [comment]
they switched to a different numbering, the latest is `2017.6`. `4.4.x` is 2016-ish, AFAIK. So yes, `2.2` is a dinosaur :smile:

**Comment by drwells** [comment]
3.0 came out in 2010, so this version must be at least seven years old.

We don't actually have a minimum version requirement for TBB as far as I can tell. It might be better to block anything before 3.0 (particularly if we don't have tests for these) since I don't think these ancient versions of TBB support C++11: we are going to keep getting burned by this.

**Comment by bangerth** [comment]
The problem appears to be that this very old version of the TBB's `tbb::thread_local_storage` class returns a *copy*, not a reference of the object in question. This fails for `std::vector<std::unique_ptr>` because `unique_ptr` is not copyable. Or that it tries to copy such an object somewhere else. Either way, this appears to be a bug in the TBB that they fixed at some later time since we don't seem to have that issue with later versions.

**Comment by asartori86** [comment]
thanks @bangerth for providing the patch. Sorry for my silence but I had some issues with my laptop and I was not able to cope with it

**Comment by bangerth** [comment]
@asartori86 -- I think we patched the problem since then in #4688.
